iiasa/ixmp
# Adapt to new RTD configuration style

This PR aims to get our current RTD builds in line with the upcoming changes to the required configuration file and prevent future build issues due to dependency changes at the same time.

Looking into our docs settings in more detail, I found that the upcoming changes do not just mean that *a* RTD config file v2 is required, but some options that are now specified at <project on RTD> -> Advanced Settings -> Default Settings will be removed in favor of options in the config file, which already overwrite them if such a file is present, even if they are not specified there. That should explain why no PDF builds are produced at the moment. So step 1 of this PR is to append the config file such that these options are active again.

Recently, our builds of the message_data docs have started to fail. The reason for that, as far as I can tell, is that the default pip installation process on RTD used to include `upgrade-strategy eager`, but recently changed to `upgrade-strategy only-if-needed`. This seems to be in line with [popular demand](https://github.com/readthedocs/readthedocs.org/issues/8890). Since I couldn't find any mention in the docs on how to enable the upgrade strategy we are used to again (apart from taking complete control and specifying every step of the workflow ourselves), I decided it might be best to follow RTD's recommendation for [reproducible builds](https://docs.readthedocs.io/en/latest/guides/reproducible-builds.html) by pinning the package versions we require. For internal consistency, I want to do so on ixmp, message_ix, message-ix-models, and message_data.
Following the recommendation, I performed the following steps to reach the version of `doc/requirements.txt` present here:

1. Activate venv, `python -m pip install pip-tools`
2. `pip-compile --allow-unsafe --extra docs -o doc/requirements.txt pyproject.toml`
3. Delete the lines following lines from the file manually:
```bash
ixmp @ file:///home/fridolin/ixmp
    # via ixmp (pyproject.toml)
```
Some notes on this process:

- `--allow-unsafe` is going to be the default soon and so it is recommended to [already run pip-compile](https://pip-tools.readthedocs.io/en/latest/#deprecations) with it now
- I haven't figured out how to automatically exclude the lines regarding ixmp, they seem to be too localized to work on RTD and unnecessary, too, since we are installing ixmp from the local git clone, either way. This, however, means that one cannot expect to run ixmp after installing all packages listed in `doc/requirements.txt`, since ixmp itself will be missing. The file only provides all dependencies that would be installed with `pip install ixmp[docs]` now. @khaeru, do you think it's enough that this file is hidden in `doc/` or should we add a comment at the top stating this exact limitation? 

Finally, I still have to check and figure out how/why the bulleted lists are not rendering correctly. If this issue exists here as well, as I expect, I might use this PR to add commits with related fixes. 

## How to review

- Read the diff and note that the CI checks all pass.
- Activate build on RTD for this branch and check that the builds are successful (note: formats other than HTML are too resource-expansive and will only be built [after a merge](https://docs.readthedocs.io/en/stable/config-file/v2.html#formats))
- Build the documentation and look at the bullet lists.

## PR checklist

- ~[ ] Add or expand tests; coverage checks both ✅~ Internal, docs-related, only need them to build/render correctly
- ~[ ] Add, expand, or update documentation.~ Only related to config of docs.
- ~[ ] Update release notes.~ Only related to config of docs.
